File: 08_gpu_teams/problems/00_offload_loop/offload_loop.cpp
#include <cstdio>
#include <cstdlib>
#include <omp.h>

int main(int argc, char ** argv) {
  long m = (1 < argc ? atol(argv[1]) : 8);
  // TODO: 下の for 文の直前に #pragma omp target teams distribute parallel for を1行追加し, ループをGPU上の多数のチーム×スレッドで並列実行させよ. (結果を表示するだけなので map 節は不要)
  for (long i = 0; i < m; i++) {
    printf("i = %ld  executed by team %d  thread %d\n",
           i, omp_get_team_num(), omp_get_thread_num());
  }
  return 0;
}

File: 08_gpu_teams/problems/00_offload_loop/offload_loop.f90
program offload_loop
  use omp_lib
  implicit none
  character(len=32) :: arg
  integer(8) :: m, i

  if (command_argument_count() >= 1) then
     call get_command_argument(1, arg); read (arg, *) m
  else
     m = 8
  end if
  ! TODO: 下の do ループを !$omp target teams distribute parallel do ... !$omp end target teams distribute parallel do で囲み, ループをGPU上の多数のチーム×スレッドで並列実行させよ. (結果を表示するだけなので map 節は不要)
  do i = 1, m
     print "(a,i0,a,i0,a,i0)", "i = ", i, "  executed by team ", &
          omp_get_team_num(), "  thread ", omp_get_thread_num()
  end do
  ! TODO: 上で始めた target teams distribute parallel do 領域を閉じる (!$omp end target teams distribute parallel do).
end program offload_loop
